compiler/bindings/python/iree/compiler/tools/onnx.py

```diff
@@ -0,0 +1,165 @@
+import os
+import tempfile
+import logging
+import onnx
+
+from dataclasses import dataclass
+from typing import IO, Any, Optional, Union
+from pathlib import Path
+
+from .. import CompilerOptions, InputType, TempFileSaver
+from ..extras import onnx_importer
+from ..ir import Context, StringAttr
+from .binaries import invoke_pipeline
+from .core import build_compile_command_line
+from ...runtime import Config, VmInstance, VmModule, load_vm_module
+
+__all__ = [
+    "compile_saved_model",
+    "compile_model"
+]
+
+
+logger = logging.getLogger(__name__)
+
+
+def load_onnx_model(raw_model: onnx.ModelProto, 
+                    data_prop: bool = True, 
+                    data_dir : IO[bytes] | str | os.PathLike | None = None):
+    # Load the model, with possible external data coming from the default
+    # location, or the location specified on the command line.
+    if data_dir:
+        onnx.load_external_data_for_model(raw_model, str(data_dir))
+
+    # Do shape inference two ways.  First, attempt in-memory to avoid redundant
+    # loading and the need for writing a temporary file somewhere.  If that
+    # fails, typically because of the 2 GB protobuf size limit, try again via
+    # files.  See
+    # https://onnx.ai/onnx/repo-docs/PythonAPIOverview.html#shape-inference-a-large-onnx-model-2gb
+    # for details about the file-based technique.
+
+    # Run the checker to test whether the file is above the threshold for
+    # in-memory shape inference.  If not, go ahead and do the shape inference.
+    try:
+        onnx.checker.check_model(raw_model)
+        inferred_model = onnx.shape_inference.infer_shapes(
+            raw_model, data_prop=data_prop
+        )
+        return inferred_model
+    except ValueError:
+        pass
```

Comment:
I'd like to see unit tests for each of these branches in the code. Hard to tell from just looking at this code if this fallthrough is acceptable or an overlooked error condition.

Reply:
How do we want to test this. I assume this would require a model larger than 2GB and we probably want to avoid checking in such a large model. I've been generating larger models in memory but the peak RAM usage is way more than 2GB for my current approach.

```python
import numpy as np

from onnx import TensorProto, save_model
from onnx.helper import (
    make_model, make_node, make_graph, make_tensor,
    make_tensor_value_info, np_dtype_to_tensor_dtype)
from onnx.checker import check_model

N = 25600

# inputs
X = make_tensor_value_info('X', TensorProto.FLOAT, [N, N])
Y = make_tensor_value_info('Y', TensorProto.FLOAT, [N, N])

# outputs
Y = make_tensor_value_info('Y', TensorProto.FLOAT, [N, N])

# nodes
node1 = make_node('MatMul', ['X', 'W'], ['Y'])

# weights
data = np.random.randn(N, N).astype(np.float32)
weight_init = make_tensor(
            name='W',
            data_type=np_dtype_to_tensor_dtype(data.dtype),
            dims=data.shape,
            vals=data)

graph = make_graph([node1],
    'test-model',
    [X],
    [Y],
    initializer=[weight_init]
)


# onnx graph
onnx_model = make_model(graph)
onnx_model.opset_import[0].version = 17

# check_model(onnx_model)

# print(onnx_model)

save_model(onnx_model, 'test.onnx')
```